docs/specs/balance-fetcher.md

```diff
@@ -0,0 +1,311 @@
+# Balance Fetcher Technical Specification
+
+## 1. Overview
+
+The Balance Fetcher is a critical component in the AEP Fee Calculator system that retrieves and stores historical balance information for reward distributors on the Arbitrum Nova network. It queries blockchain state at specific end-of-day blocks to capture point-in-time balance snapshots necessary for accurate fee calculations.
+
+## 2. Purpose
+
+### Primary Objectives
+
+- Retrieve end-of-day ETH balances for all tracked reward distributors
+- Store historical balance data in a consistent, accessible format
+- Support incremental processing to handle new dates efficiently
+- Handle distributors created mid-period by only fetching balances from their creation date onward
+
+### Component Responsibilities
+
+- Query blockchain state using RPC calls to retrieve balances
+- Manage balance data persistence for each distributor
+- Ensure data integrity and handle RPC failures gracefully
+- Maintain idempotent operations for reliability
+
+### Success Criteria
+
+- All tracked distributors have balance data for all applicable dates
+- Balance data accurately reflects on-chain state at specified blocks
+- Component can process new dates incrementally without reprocessing existing data
+- Balance fetching respects distributor creation dates (no queries before creation)
+
+## 3. Dependencies
+
+### External Libraries
+
+- `ethers` (v6.x) - For RPC communication and blockchain interaction
+- Node.js built-in modules: `fs/promises`, `path`
+
+### System Components
+
+- **FileManager**: For reading and writing balance data files
+- **RPC Provider**: Arbitrum Nova RPC endpoint for blockchain queries
+
+### Infrastructure Requirements
+
+- Reliable RPC endpoint with support for `eth_getBalance` at historical blocks
+- File system access for reading distributor list and block numbers
+- Write permissions for storing balance data
+
+## 4. Data Flow
+
+### Input Flow
+
+1. Read distributor addresses from `store/distributors.json`
+2. For each distributor:
+   - Read master block numbers from `store/block_numbers.json`
+   - Load existing balance data from `store/distributors/{address}/balances.json` (if exists)
+   - Determine which dates need balance queries
+
+### Process Flow
+
+1. Initialize RPC provider connection
+2. For each distributor address:
+   - Load distributor metadata to get creation date
+   - Filter block numbers to only include dates from creation onward
+   - For each applicable date without existing balance data:
+     - Query balance using `eth_getBalance(address, blockNumber)`
+     - Store result in memory
+   - Write updated balance data to distributor's balance file
+
+### Output Flow
+
+- Balance data written to `store/distributors/{address}/balances.json`
```

Comment:
All updates to file should be done via the file manager - update this doc everyhwere